### ISSUE TITLE
Using `done` callbacks when chai-as-promised is present leads to horrible silent errors

### ISSUE THREAD

**Issue by mmomtchev** [chaijs/chai#1402]
I don't know if this is supported or not, but it causes a very vicious silent error which causes the rest of the unit tests to not run and **mocha** to immediately exit with 0, reporting a global success - which means that automated unit testing doesn't report any error

It is somewhat random, depending on the Node version - I have a repo that can reproduce it with 100% probability on Node 14 on Ubuntu 20.04

(`ds.root.arrays.getAsync('time')` returns a Promise that resolves with a `gdal.MDArray`)

```js
it('should have "getAsync()" method', (done) => {
  assert.eventually.instanceOf(ds.root.arrays.getAsync('time'), gdal.MDArray).then(done)
})
```

It never happens when the test is run as a single test

Transforming it to:
```js
it('should have "getAsync()" method', () =>
  assert.eventually.instanceOf(ds.root.arrays.getAsync('time'), gdal.MDArray)
)
```
triggers (sometimes, randomly) the `afterEach` hook before the Promise has been resolved

Here is the repo that fails 100% of the time with Node 14 on Ubuntu 20.04
```
git clone https://github.com/mmomtchev/node-gdal-async.git -b v3.2.991
npm install
mocha
```

**Comment by mmomtchev** [comment]
I understood the underlying problem that was causing it, I will try to make a simple repro

**Comment by mmomtchev** [comment]
Ok, I leave it to you how to handle this - it is an error that is impossible to produce using only JS.

Basically `getAsync` is a C++ method with a callback. It encounters an error and then places an exception on the stack, but then doesn't handle correctly the error condition and calls the JS callback - which in this case is provided by Node.js `promisify`. The exception stays on the stack and it is triggered in a very unusual way and sometimes propagates up one level more than it should - depending on the exact V8 version.

**Comment by sukrosono** [comment]
Hi thanks for the issue, please include the repro. It might the internal team want to dive right away when this is not intended. Or this just specific to node v.14. You also could provide more detail about your environment, like chai and mocha version.

**Comment by mmomtchev** [comment]
@sukrosono Thanks for your reply, I can probably try to make a small Node.js addon that does this, all that is needed is a C++ function that accepts an `(e, r)` callback and does:
```cpp
Nan::ThrowError("error");
callback->Call(2, { Nan::Null(), Nan::Null() }, async_resource);
```
and then use the above construct in chai/mocha with a `done` - mocha will immediately exit and report a success.

The problem is that, from mocha/chai standpoint, this is an abnormal V8 behavior since, by returning with a pending exception, the user program is able to find a way out of its enclosing `try...catch` block - which is shouldn't be possible according to the JS specifications.

**Comment by mmomtchev** [comment]
@sukrosono I think I found the problem: https://github.com/nodejs/node/issues/29355

**Comment by jdmarshall** [comment]
Does that mean this issue can be closed?

**Comment by mmomtchev** [comment]
No, it means that the problem is understood. It is especially in mocha that this behaviour is problematic since it can hide errors.
